### ISSUE TITLE
ccall, cwrap non-ASCII string handling still ain't no good

### ISSUE THREAD

**Issue by H2CO3** [emscripten-core/emscripten#2752]
**Bug:** Although `writeStringToMemory()` correctly processes the string and makes it into a UTF-8 array of the appropriate length, `stringToC()` still makes the wrong assumption that each character will fit into 1 byte. Thus, `writeStringToMemory()` will presumably write out-of-bounds and overflow the `ret` buffer.

**Reference in source:** [preamble.js @ line #360](https://github.com/kripken/emscripten/blob/206b05b7e8c48c71f2dcb360b7aba1887410af70/src/preamble.js#L360)

**Proposed fix:** since in UTF-8, every character (code point?) is representable by at most 4 bytes, I suggest allocating 4 bytes for each character:

```
ret = Runtime.stackAlloc(str.length * 4 + 1); // +1 for the trailing '\0'
```

Unless a string is enormously large, this should not cause any problems.

**Comment by H2CO3** [comment]
Nice. `wontfix` a memory corruption bug. Good job. 👍

**Comment by kripken** [comment]
@H2CO3 the `wontfix` label was added automatically by a bot. It was also automatically removed once a comment was added to the issue (as it's no longer stale).

Anyhow this appears to have been fixed: https://github.com/emscripten-core/emscripten/blob/incoming/src/preamble.js#L88 and testing locally it does look like it works safely, so I think we can close this.

If I missed something here, please reopen.

**Comment by sbc100** [comment]
> Nice. `wontfix` a memory corruption bug. Good job.

Sorry for confusion there.  The spirit in which the bot is marking these tickets is good i assure your.  Perhaps the label should have been "stale" .. and then "wontfix" only after the 2 week period.

Some rational for the general approach: https://github.com/probot/stale#is-closing-stale-issues-really-a-good-idea
